src/osd/PrimaryLogPG.cc

```diff
@@ -2587,11 +2587,12 @@ int PrimaryLogPG::do_manifest_flush(OpRequestRef op, ObjectContextRef obc, Flush
 	}
       }();
       bufferlist in;
-      if (fp_oid != tgt_soid.oid) {
+      if (fp_oid != tgt_soid.oid && (!obc->ssc && !obc->ssc->snapset.clones.size())) {
```

Comment:
I think you probably want to revert this line.  I'm not sure what (!obc->ssc && !obc->ssc->snapset.clones.size()) is meant to do, but:
1. !obc->ssc implies that obc->ssc->snapset is undefined behavior, so this cannot be right
2. forall obc, obc->ssc is non-null.

Sorry for not noticing this earlier.

Reply:
We don't need to decrement the refcount in the case of snapshotted manifest object. See below case.
snap[0]: [bb] 
head:     [er]
bb's refoucnt should be 1, hi's refcount should be 1 (assuming that both bb and hi have the same offset).
But, in above case, bb's refcount is decreased without (!obc->ssc && !obc->ssc->snapset.clones.size()) when flush.
So, what (!obc->ssc && !obc->ssc->snapset.clones.size()) means that do not decrement the refcount if this is snapshot.

Also, I'm preparing a next PR related with this 
(https://github.com/myoungwon/ceph/commits/wip-refcount-snap-before-flush).
We need to avoid incrementing the refcount in the following case when flush.
snap[0]: [bb] [hi]
head:     [er]  [hi]
hi's refcount should be 1.

Reply:
In that case, you need to actually fix the condition.  Reread my comment.  As written, that branch will *never* run because obc->ssc is always populated.

Reply:
Nvm, looks like you fixed it.